tests/cpp/test_multidevice_transformer.cpp

```diff
@@ -278,29 +278,65 @@ std::vector<TensorView*> mlp(
     TensorView* b0,
     TensorView* w1,
     TensorView* b1,
-    const DeviceMesh& mesh) {
+    const DeviceMesh& mesh,
+    bool sequence_parallel = false,
+    TensorView* mask = nullptr) {
   const DataType dtype = w0->dtype();
-  // // Linear 0
+
+  if (sequence_parallel) {
+    // Input arrives sharded and must be allgathered back
+    x->setDeviceMesh(mesh);
+    x->axis(0)->parallelize(ParallelType::DIDx);
+    x = set(x); // allgather
+    x->setDeviceMesh(mesh);
+    x->axis(0)->parallelize(ParallelType::Serial);
+    // Reshape back to 2D for linearOp
+    auto D = w0->axis(0)->extent()->value().as<int64_t>();
+    x = reshape(x, {D, B * S / D, E}, {B * S, E});
+  }
+  // Linear 0
   TensorView* linear0 = linear(x, w0, b0);
   // GeLU
   TensorView* gelu = tanh_gelu(castOp(DataType::Float, linear0));
   gelu = castOp(dtype, gelu);
   // Linear 1
   TensorView* local_matmul1 = matmul(gelu, transpose(w1, 1, 2));
-  TensorView* matmul1 = sum(local_matmul1, {0}); // Allreduce
-  TensorView* linear1 = add(matmul1, broadcast(b1, {true, false}));
+  if (sequence_parallel) {
+    // Remove after https://github.com/NVIDIA/Fuser/issues/2563
+    // Reshape to explicitly pull the sharded axis into the logical domain
+    auto D = w0->axis(0)->extent()->value().as<int64_t>();
+    local_matmul1 = reshape(local_matmul1, {D, B * S, E}, {D, D, B * S / D, E});
+  }
+  TensorView* matmul1 = sum(local_matmul1, {0}); // Allreduce or Reduce scatter
+  TensorView* linear1 = add(matmul1, broadcast(b1, {true, true, false}));
   // Dropout
   Val* prob = IrBuilder::create<Val>(1.0 - kDropoutProb);
   Val* scale = IrBuilder::create<Val>(1.0 / (1.0 - kDropoutProb));
-  auto dropout_result = dropout(linear1, prob, scale).output;
+  if (mask == nullptr) {
+    auto rand_vals = rand_like(linear1);
```

Comment:
(not required for this PR) 

Should we also feed seed and offset? rand_like can already take that. It's in general better to expose rng outside so it's easier to make things deterministic yet random. I think this is helpful for implementing sharded dropout where each rank has to be randomized differently.

Reply:
Yes, that would be better. I did attempt setting the offset and seed in the nvfuser fusion, but I couldn't figure out exactly how to set the parameters to match ATen's dropout. 

I assumed that ATen's dropout would advance the offset by the size of the input, but it only advanceded the offset by 1 for each call to dropout. 
Some pytorch code showing this
```
>>> x = torch.rand([5])
tensor([0.4963, 0.7682, 0.0885, 0.1320, 0.3074])
>>> torch.nn.functional.dropout(x)
tensor([0.0000, 1.5364, 0.1770, 0.2641, 0.6148])
>>> torch.rand([5])
tensor([0.4901, 0.8964, 0.4556, 0.6323, 0.3489])

>>> torch.manual_seed(0)
<torch._C.Generator object at 0x7f819c426830>
>>> torch.rand([10])
tensor([0.4963, 0.7682, 0.0885, 0.1320, 0.3074, 0.6341, 0.4901, 0.8964, 0.4556,
        0.6323])
The dropout only caused 0.6341 to be skipped
  ```

Reply:
Got it! I saw Thunder solve this by using a lower-level op: https://github.com/Lightning-AI/lightning-thunder/blob/1575797a8d21e0b845e188ddd6af305be6457b1c/thunder/executors/torchex.py#L187. Certainly not ideal because have to manually decompose. But worth noting when we really need this.